rpc/types/utils.go

```diff
@@ -184,22 +189,30 @@ func NewRPCTransaction(
 	} else {
 		signer = ethtypes.HomesteadSigner{}
 	}
-	from, _ := ethtypes.Sender(signer, tx) // #nosec G703
+	originalFrom, _ := ethtypes.Sender(signer, tx) // #nosec G703
+
+	from := originalFrom
+	if onBehalf != nil {
+		// Replace the sender address if the transaction is impersonated
+		from = *onBehalf
+	}
```

Comment:
this part is important as it replaces `from` in all the `eth_getTransactionBy*` RPC calls.